dask/dataframe/tests/test_dataframe.py

```diff
@@ -382,25 +324,108 @@ def test_arithmetic_frame():
+        check_frame_arithmetics(l, r, el, er)
+
+    # different index, pandas raises ValueError in comparison ops
+
+    dsk5 = pd.DataFrame({'a': [3, 2, 1, 5, 2, 8, 1, 4, 10],
+                         'b': [7, 8, 9, 4, 2, 3, 1, 0, 5]},
+                        index=[0, 1, 3, 5, 6, 8, 9, 9, 9])
+    d5 = dd.from_pandas(dsk5, 2)
+    full5 = d5.compute()
+    dsk6 = pd.DataFrame({'a': [3, 2, 1, 5, 2 ,8, 1, 4, 10],
+                         'b': [7, 8, 9, 5, 7, 8, 4, 2, 5]},
+                        index=[0, 1, 2, 3, 4, 5, 6, 7, 9])
+    d6 = dd.from_pandas(dsk6, 4)
+    full6 = d6.compute()
+
+    full7 = pd.DataFrame({'a': [1, 2, 3, 4, 5, 6, 7, 8],
+                          'b': [5, 6, 7, 8, 1, 2, 3, 4]},
+                         index=list('aaabcdeh'))
+    full8 = pd.DataFrame({'a': [5, 6, 7, 8, 4, 3, 2, 1],
+                          'b': [2, 4, 5, 3, 4, 2, 1, 0]},
+                         index=list('abcdefgh'))
+    d7 = dd.from_pandas(full7, 3)
+    d8 = dd.from_pandas(full8, 4)
+
+    # Arithmetics with different index
+    cases = [(d5, d6, full5, full6),
+             (d5.repartition([0, 9]), d6, full5, full6),
+             (d5.repartition([0, 5, 9]), d6.repartition([0, 7, 9]), full5, full6),
+             (d7, d8, full7, full8),
+             (d7.repartition(['a', 'c', 'h']), d8.repartition(['a', 'h']),
+              full7, full8),
+             (d7.repartition(['a', 'b', 'e', 'h']), d8.repartition(['a', 'e', 'h']),
+              full7, full8)]
 
-        assert eq(l + r, el + er)
-        assert eq(l * r, el * er)
-        assert eq(l - r, el - er)
-        assert eq(l / r, el / er)
+    for (l, r, el, er) in cases:
+        check_series_arithmetics(l.a, r.b, el.a, er.b, allow_comparison_ops=False)
+        check_frame_arithmetics(l, r, el, er, allow_comparison_ops=False)
+
+    # Not (yet) supported iff left/right edge of divisions are different
+    full7 = pd.DataFrame({'a': [1, 2, 3, 4, 5, 6, 7, 8],
+                          'b': [5, 6, 7, 8, 1, 2, 3, 4]},
+                         index=[0, 2, 4, 8, 9, 10, 11, 13])
+    full8 = pd.DataFrame({'a': [5, 6, 7, 8, 4, 3, 2, 1],
+                          'b': [2, 4, 5, 3, 4, 2, 1, 0]},
+                         index=[1, 3, 4, 8, 9, 11, 12, 13])
+    d7 = dd.from_pandas(full7, 3)
+    d8 = dd.from_pandas(full8, 2)
+
+    full9 = pd.DataFrame({'a': [1, 2, 3, 4, 5, 6, 7, 8],
+                          'b': [5, 6, 7, 8, 1, 2, 3, 4]},
+                         index=[0, 2, 4, 8, 9, 10, 11, 13])
+    full10 = pd.DataFrame({'a': [5, 6, 7, 8, 4, 3, 2, 1],
+                          'b': [2, 4, 5, 3, 4, 2, 1, 0]},
+                         index=[0, 3, 4, 8, 9, 11, 12, 12])
+    d9 = dd.from_pandas(full9, 3)
+    d10 = dd.from_pandas(full10, 2)
+
+    assert raises(ValueError, lambda: d7 + d8)
+    assert raises(ValueError, lambda: d7.a + d8.b)
+    assert raises(ValueError, lambda: d9 + d10)
+    assert raises(ValueError, lambda: d9.a + d10.b)
```

Comment:
Currently it is validated explicitly, so it should raise.

- Before this PR: Check all divisions are identical
- After this PR: Check only left-most / right-most edge of divisions are the same